### PR TITLE
[BALDE-23] Window abstraction

### PR DESCRIPTION
**Pull request by 4jeR**
Window abstaction

### CODE REVIEW

File: src/graphics/Window.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+         */
+        constexpr void Update();
+
+        /**
+         * @fn                      IsClosed
+         * @brief                   updates the window data
+         * @return [int]            true  -> Window is     closed
+         *                          false -> Window is NOT closed
+         */
+        [[nodiscard]] inline int IsClosed() const noexcept { return glfwWindowShouldClose(m_Window); }
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Shouldn't return m_Closed?

File: src/graphics/Window.h [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        /**
+         * @fn                      Window
+         * @param [std::string]     title
+         * @param [int]             width
+         * @param [int]             height
+         * @brief                   constructor for the window
+         */
+        explicit Window(const char *title = "Window", int width = 800, int height = 600,
+                        bool closed = false)
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Shouldn't closed be true by default? When ::Init will be invoked and everything will be run correctly than we can certainly tell that window isNotClosed.

File: src/graphics/Window.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        /**
+         * @fn                      Update
+         * @brief                   updates the window data i.a. poll events & swap buffers
+         */
+        constexpr void Update();
+
+        /**
+         * @fn                      IsClosed
+         * @brief                   updates the window data
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Does it update anything? Or it just returns whether window is closed or not?

File: src/graphics/Window.h [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+         *                          false -> Window is NOT closed
+         */
+        [[nodiscard]] inline int IsClosed() const noexcept { return glfwWindowShouldClose(m_Window); }
+
+        /**
+         * @fn                      DestroyWindow
+         * @brief                   destroys the Window
+         */
+        void DestroyWindow() const noexcept {
+            glfwDestroyWindow(m_Window);
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Shouldn't destroy make the window closed?

**Review by piotrmoszkowicz — CHANGES_REQUESTED** [review]

File: src/graphics/Window.cpp [hunk 5]
```diff
@@ -6,51 +6,33 @@
...
             glfwPollEvents();
+            glfwGetFramebufferSize(m_Window, &m_Width, &m_Height);
+            glViewport(0, 0, m_Width, m_Height);
             glfwSwapBuffers(m_Window);
         }
-
     }
 
-
     bool Window::Init() noexcept {
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.cpp]: Is it OK to re-init same window one more time?
Maybe there's if(!m_Closed) check needed?

**Review by piotrmoszkowicz — CHANGES_REQUESTED** [review]

File: src/graphics/Window.h (hunk 1 above)
Comment (4jeR) [reply src/graphics/Window.h]: Actually I was wondering about it, probably You are right

File: src/graphics/Window.h (hunk 3 above)
Comment (4jeR) [reply src/graphics/Window.h]: copy pastee yooo :((( :dancer:

File: src/graphics/Window.cpp (hunk 5 above)
Comment (BIGbadEL) [reply src/graphics/Window.cpp]: maybe just make it private

**Review by BIGbadEL — COMMENTED** [review]
you should get rid of all constexpr, because it's impossible to call any of that functions at compile time

File: src/graphics/Window.h (hunk 4 above)
Comment (4jeR) [reply src/graphics/Window.h]: As far as I know this is how it's done, will add additional m_Closed = true;
Correct me if needed though

File: src/graphics/Window.h (hunk 2 above)
Comment (4jeR) [reply src/graphics/Window.h]: fixed

File: src/graphics/Window.cpp (hunk 5 above)
Comment (4jeR) [reply src/graphics/Window.cpp]: Method **Init()** is private indeed

File: src/graphics/Window.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
+//
+// Created by ajer on 11.02.19.
+//
+
+#include "Window.h"
+
+namespace Bald::Graphics {
+
+    void Window::Update() noexcept {
+        if (!m_Closed) {
```
Comment (Blinkuu) [inline src/graphics/Window.cpp]: This shouldn't be checked here. Update is update.

File: src/graphics/Window.cpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        }
+    }
+
+    bool Window::Init() noexcept {
+
+        if (!m_Closed) {
+            CORE_LOG_WARN("[Window] Window already opened!"); // or CORE_LOG_ERROR ?
+            return false;
+        }
+        /**Not sure about this ^ */
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.cpp]: Indeed not needed.

File: src/graphics/Window.h [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+         * @brief                   destructor - terminates the m_Window
+         */
+        ~Window() noexcept { glfwTerminate(); } /** shouldn't be glfwDestroyWindow() ? */
+
+
+        /**
+         * @fn                      Update
+         * @brief                   updates the window data i.a. poll events & swap buffers
+         */
+        void Update()noexcept;
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Space

File: src/graphics/Window.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+
+        /** ========== GETTERS ========== */
+
+        /**
+         * @fn                      GetWindow
+         * @brief                   returns m_Window component
+         * @return [GLFWwindow*]    pointer to the m_Window
+         */
+        [[nodiscard]]  inline GLFWwindow *GetWindow() const noexcept { return m_Window; }
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Remove space

File: src/graphics/Window.h [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+        /**
+         * @fn                      GetTitle
+         * @brief                   returns m_Title component
+         * @return [const char*]    pointer to the m_Title
+         */
+        [[nodiscard]]  inline const char *GetTitle() const noexcept { return m_Title; }
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Remove space

File: src/graphics/Window.h [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+        /**
+         * @fn                      GetWidth
+         * @brief                   returns the width of the Window
+         * @return [int]            m_Width component
+         */
+        [[nodiscard]]  inline int GetWidth() const noexcept { return m_Width; }
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Remove space

File: src/graphics/Window.h [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+        /**
+         * @fn                      GetHeight
+         * @brief                   returns the height of the Window
+         * @return [int]            m_Height component
+         */
+        [[nodiscard]]  inline int GetHeight() const noexcept { return m_Height; }
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: Remove space

File: src/graphics/Window.h [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+         */
+        bool Init() noexcept;
+
+        // START OF ATTRIBUTES
+
+    private:
+        GLFWwindow* m_Window;
+        const char* m_Title;
+        int m_Width;        //
+        int m_Height;       // should these two be of type GLInt ?
```
Comment (piotrmoszkowicz) [inline src/graphics/Window.h]: It's OK

**Review by piotrmoszkowicz — DISMISSED** [review]

File: CMakeLists.txt [hunk 14]
```diff
@@ -37,11 +38,14 @@ add_compile_options(
...
 add_library(bald_engine
-        library.cpp library.h
+        library.cpp
+        library.h
         src/utils/file_manager.cpp
         src/utils/split_string.cpp
         src/input/input_manager.cpp
-        src/Log.cpp)
+        src/Log.cpp
+        src/graphics/Window.h
```
Comment (Blinkuu) [inline CMakeLists.txt]: Remove Window.h, it is already included in line 11!

File: src/graphics/Window.cpp [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace Bald::Graphics {
+
+
+    Window::Window(const char *title, int width, int height, bool closed, bool VSync)
+            : m_Title(title), m_Width(width), m_Height(height), m_Closed(closed), m_VSync(VSync) {
+        Window::Init();
+    }
+
+    void Window::Update() noexcept {
```
Comment (Blinkuu) [inline src/graphics/Window.cpp]: Update should ONLY PollEvents and SwapBuffers!!!! Viewport should be changed once per FrameBufferCallback.

File: src/graphics/Window.cpp [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+            exit(1);
+        }
+
+        if (!glfwInit()) {
+            CORE_LOG_ERROR("[Window] Failed to init GLFW!");
+            exit(1);
+        }
+
+
+        m_Window = glfwCreateWindow(m_Width, m_Height, m_Title, NULL, NULL);
```
Comment (Blinkuu) [inline src/graphics/Window.cpp]: Use nullptr, it is safer, and we are in C++ 💃

File: src/graphics/Window.cpp [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+        if (!m_Window) {
+            glfwTerminate();
+            CORE_LOG_ERROR("[Window] Failed to create a Window!");
+            return false;
+        }
+
+        m_Closed = false;
```
Comment (Blinkuu) [inline src/graphics/Window.cpp]: Get rid of this m_Closed thing, it is not needed -> glfw provides us with a function that check whether window should be closed or not.

File: src/graphics/Window.h [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+        bool m_Closed;
```
Comment (Blinkuu) [inline src/graphics/Window.h]: This should be gone

File: src/graphics/Window.h [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+         * @brief                   constructor for the window
+         */
+        explicit Window(const char* title = "Window", int width = 800, int height = 600,
+                        bool closed = true, bool VSync = false);
+
+        /**
+         * @fn                      ~Window
+         * @brief                   destructor - terminates the m_Window
+         */
+        ~Window() noexcept { Shutdown(); }
```
Comment (Blinkuu) [inline src/graphics/Window.h]: I think it should be implemented in a .cpp file

File: src/graphics/Window.h [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+         * @return [bool]           true  -> Window is     closed
+         *                          false -> Window is NOT closed
+         */
+        [[nodiscard]] inline bool IsClosed() const noexcept { return m_Closed; }
+
+        /**
+         * @fn                      DestroyWindow
+         * @brief                   destroys the Window
+         */
+        void DestroyWindow() noexcept;
```
Comment (Blinkuu) [inline src/graphics/Window.h]: I think you shouldn't be allowed to destroy Window => leave it for Destructor which calls Shutdown();

File: src/graphics/Window.h [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+         */
+        void Update() noexcept;
+
+        /**
+         * @fn                      IsClosed
+         * @brief                   returns the m_Closed component
+         * @return [bool]           true  -> Window is     closed
+         *                          false -> Window is NOT closed
+         */
+        [[nodiscard]] inline bool IsClosed() const noexcept { return m_Closed; }
```
Comment (Blinkuu) [inline src/graphics/Window.h]: Get rid of IsClosed()

File: src/graphics/Window.h [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+         * @fn                      DestroyWindow
+         * @brief                   destroys the Window
+         */
+        void DestroyWindow() noexcept;
+
+        /**
+         * @fn                      ClearWindow
+         * @brief                   clears buffer bit
+         */
+        void ClearWindow() const noexcept;
```
Comment (Blinkuu) [inline src/graphics/Window.h]: I think it would be better to name this method just "Clear()"

File: src/graphics/Window.h [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+
+        /**
+         * @fn                      Shutdown
+         * @brief                   destroys the window
+         */
+        void Shutdown();
```
Comment (Blinkuu) [inline src/graphics/Window.h]: Should be private!

**Review by Blinkuu — CHANGES_REQUESTED** [review]

File: src/graphics/Window.cpp (hunk 6 above)
Comment (4jeR) [reply src/graphics/Window.cpp]: fixed

File: CMakeLists.txt (hunk 14 above)
Comment (4jeR) [reply CMakeLists.txt]: fixed

File: src/graphics/Window.cpp (hunk 15 above)
Comment (4jeR) [reply src/graphics/Window.cpp]: done

File: src/graphics/Window.cpp (hunk 16 above)
Comment (4jeR) [reply src/graphics/Window.cpp]: done  :dancer:

File: src/graphics/Window.cpp (hunk 17 above)
Comment (4jeR) [reply src/graphics/Window.cpp]: removed

File: src/graphics/Window.h (hunk 19 above)
Comment (4jeR) [reply src/graphics/Window.h]: moved

File: src/graphics/Window.h (hunk 20 above)
Comment (4jeR) [reply src/graphics/Window.h]: changed

File: src/graphics/Window.h (hunk 21 above)
Comment (4jeR) [reply src/graphics/Window.h]: removed

File: src/graphics/Window.h (hunk 22 above)
Comment (4jeR) [reply src/graphics/Window.h]: indeed, changed

File: src/graphics/Window.h (hunk 23 above)
Comment (4jeR) [reply src/graphics/Window.h]: done

File: src/graphics/Window.cpp [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+//
+
+#include "Window.h"
+
+#include <glad/glad.h>
+#include <GLFW/glfw3.h>
+
+namespace Bald::Graphics {
+
+    Window::Window(GLFWwindow* glfwwindow, const char* title, int width, int height, bool VSync):
```
Comment (BIGbadEL) [inline src/graphics/Window.cpp]: why do we need GLFWwindow* glfwwindow as a constructor parameter?

File: src/graphics/Window.cpp [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+        if (!gladLoadGLLoader(reinterpret_cast<GLADloadproc>(glfwGetProcAddress))) {
+            CORE_LOG_ERROR("[Window] Failed to load GLAD!");
+            exit(1);
+        }
+
+        if (!glfwInit()) {
+            CORE_LOG_ERROR("[Window] Failed to init GLFW!");
+            exit(1);
+        }
+
```
Comment (BIGbadEL) [inline src/graphics/Window.cpp]: after initializing all OpenGL we should print all the information about it and about computer we currently running

File: src/graphics/Window.cpp [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+            glfwSwapInterval(0);
+
+        m_VSync = enabled;
+    }
+
+
+    bool Window::Init() noexcept {
+
+
+        if (!gladLoadGLLoader(reinterpret_cast<GLADloadproc>(glfwGetProcAddress))) {
```
Comment (BIGbadEL) [inline src/graphics/Window.cpp]: I think you should call glfwInit, create a window and then call gladLoadGLLoader

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

File: src/graphics/Window.cpp [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+
+
+namespace Bald::Graphics {
+
+    Window::Window(const char* title, int width, int height, bool VSync) :
+            m_Title(title),
+            m_Width(width),
+            m_Height(height),
+            m_VSync(VSync) {
+        Window::Init();
```
Comment (BIGbadEL) [inline src/graphics/Window.cpp]: Window::Init() would do exactly the same what Init() and init returns bool it would be smart to check what it returned

**Review by BIGbadEL — CHANGES_REQUESTED** [review]
You should add an example of using this class like input_example

File: src/graphics/Window.cpp (hunk 27 above)
Comment (Blinkuu) [reply src/graphics/Window.cpp]: I would make Init() return void and actually assert when something fails in initialization.

File: src/graphics/Window.cpp (hunk 25 above)
Comment (BIGbadEL) [reply src/graphics/Window.cpp]: !!!

File: src/graphics/Window.cpp [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+
+        glfwInit();
+
+        m_Window = glfwCreateWindow(m_Width, m_Height, m_Title, nullptr, nullptr);
+        glfwMakeContextCurrent(m_Window);
+        if (!gladLoadGLLoader(reinterpret_cast<GLADloadproc>(glfwGetProcAddress))) {
+            CORE_LOG_ERROR("[Window] Failed to load GLAD!");
+            exit(1);
+        }
+
```
Comment (BIGbadEL) [inline src/graphics/Window.cpp]: input callbacks should be set here

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

**Review by Blinkuu — APPROVED** [review]

File: src/input/input_manager.h [hunk 29]
```diff
@@ -309,9 +347,7 @@ namespace Bald::Input {
 
     template<class F, class... Args>
     void InputManager::SetMouseScrolledCallback([[maybe_unused]]F&& fun, [[maybe_unused]]Args&& ... args) noexcept {
```
Comment (BIGbadEL) [inline src/input/input_manager.h]: [[maybe_unused]] can be removed

**Review by BIGbadEL — DISMISSED** [review]

**Pull request merged**